### PR TITLE
MindSpore Lite support

### PR DESCRIPTION
**Pull request by YvetteLaw**
Hello, I try to make Netron support MindSpore Lite (.ms). Please check if my work meets your requirements. If there is any problem, I will revise it. What's more, I still need to make some changes, please do not merge for now.

See #600

### CODE REVIEW

**Comment by lutzroeder** [comment]
[ ] add test files to `./test/models.json`

File: source/view.js [hunk 1]
```diff
@@ -1245,6 +1245,7 @@ view.ModelFactoryService = class {
         this.register('./npz', [ '.npz', '.h5', '.hd5', '.hdf5' ]);
         this.register('./dl4j', [ '.zip' ]);
         this.register('./mlnet', [ '.zip' ]);
+        this.register('./mslite', [ '.ms']);
```
Comment (lutzroeder) [inline source/view.js:1248]: This is confusing. `.ms` is already used as an extension by other formats. The FlatBuffer has no `file_identifier` which makes it impossible to distinguish these formats. 

- [ ] can the file extension be changed to `.mslite` to be consistent with the module name?
- [ ] can a `file_identifier` be added to `schema.fbs`?

File: tools/mslite [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+clean() {
+    bold "mindspore clean"
+    rm -rf "./third_party/source/mindspore"
+}
+
+sync() {
+    bold "mindspore sync"
+    [ -d "./third_party/source/mindspore" ] || git clone --quiet https://github.com/mindspore-ai/mindspore.git "./third_party/source/mindspore"
```
Comment (lutzroeder) [inline tools/mslite:17]: Instead of cloning the entire repo only `curl` the `.fbs` files. See `./tools/paddle`.

File: tools/mslite_typeid.txt [hunk 3]
```diff
@@ -0,0 +1,48 @@
+$root.mslite.TypeId = {
```
Comment (lutzroeder) [inline tools/mslite_typeid.txt:1]: Why is this needed? This should be either part of an `.fbs` file or coded in `./source/ms.js`.

File: tools/mslite-script.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+
+def main():
```
Comment (lutzroeder) [inline tools/mslite-script.py:2]: See below. This should not be needed.

File: tools/mslite [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    git pull --quiet --prune
+    popd > /dev/null
+}
+
+schema() {
+    echo "mslite schema"
+
+    # modify
+    cp ./third_party/source/mindspore/mindspore/lite/schema/*.fbs ./tools/
+    sed -i '1,16d' ./tools/model.fbs
```
Comment (lutzroeder) [inline tools/mslite:28]: Why are schema file changes needed? Are these bugs in the schema file? Please fix and remove.

**Review by lutzroeder — CHANGES_REQUESTED** [review]

File: tools/mslite-script.py (hunk 4 above)
Comment (YvetteLaw) [reply tools/mslite-script.py:2]: We need it to process mslite-schema.js generated by flatc.js to fit the Netron framework. Currently, MindSpore Lite is used in many models and scenarios without bugs. However, if we modify its model.fbs and ops.fbs , it will have a great impact. Can we keep the file mslite-script.py?

Comment (lutzroeder) [reply tools/mslite-script.py:2]: `op.primitive.value.constructor.name` should return the same result as `op.primitive.type`.

**Comment by YvetteLaw** [comment]
https://github.com/lutzroeder/netron/pull/602#pullrequestreview-493599449

> * [ ]  add test files to `./test/models.json`
> * [ ]  revert `./setup.py` changes
> * [ ]  revert `./README.md` changes
> * [ ]  add `./source/mslite-metadata.json` with operator categories and argument names
> * [ ]  resolve issues below

I've modified files as suggested.

File: source/app.js [hunk 6]
```diff
@@ -133,7 +133,8 @@ class Application {
                     'pkl', 'joblib',
                     'pbtxt', 'prototxt',
                     'cfg',
-                    'xml' ] }
+                    'xml',
+                    'ms' ] }
```
Comment (lutzroeder) [inline source/app.js:137]: `.ms` is ambiguous. Please work with the team to change this to a more descriptive extension like `.mslite`.

File: tools/mslite [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+clean() {
+    bold "mindspore clean"
+    rm -rf "./third_party/source/mindspore"
+}
+
+sync() {
+    bold "mindspore sync"
+    mkdir -p "./third_party/source/mindspore/"
+    curl --silent --location --output "./third_party/source/mindspore/model.fbs" "https://github.com/mindspore-ai/mindspore/blob/master/mindspore/lite/schema/model.fbs?raw=true"
```
Comment (lutzroeder) [inline tools/mslite:18]: Keep the original path `./third_party/mindspore/mindspore/lite/schema/model.fbs`.

File: tools/mslite [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+clean() {
+    bold "mindspore clean"
+    rm -rf "./third_party/source/mindspore"
+}
+
+sync() {
+    bold "mindspore sync"
+    mkdir -p "./third_party/source/mindspore/"
```
Comment (lutzroeder) [inline tools/mslite:17]: Keep the original path `./third_party/mindspore/mindspore/lite/schema`.

File: tools/mslite [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    curl --silent --location --output "./third_party/source/mindspore/model.fbs" "https://github.com/mindspore-ai/mindspore/blob/master/mindspore/lite/schema/model.fbs?raw=true"
+    curl --silent --location --output "./third_party/source/mindspore/ops.fbs" "https://github.com/mindspore-ai/mindspore/blob/master/mindspore/lite/schema/ops.fbs?raw=true"
+}
+
+schema() {
+    echo "mslite schema"
+
+    # modify
+    cp ./third_party/source/mindspore/*.fbs ./tools/
+    sed -i '1,16d' ./tools/model.fbs
```
Comment (lutzroeder) [inline tools/mslite:27]: Please remove all `sed` commands and fix either the original `.fbs` schema file or `mslite.js`.

File: test/models.json [hunk 10]
```diff
@@ -5722,9 +5722,15 @@
     "format": "UFF v1"
   },
   {
-    "type":   "weka",
```
Comment (lutzroeder) [inline test/models.json:5725]: Undo spacing changes.

**Review by YvetteLaw — COMMENTED** [review]
https://github.com/lutzroeder/netron/pull/602#discussion_r504362264

Can flatc.js process long(int64) data?

**Comment by lutzroeder** [comment]
> Can flatc.js process long(int64) data?

Can you provide an example?

Fix for the `/* */` and `a.b` namespace issues is now in `master`.

**Comment by YvetteLaw** [comment]
> > Can flatc.js process long(int64) data?
> 
> Can you provide an example?
> 
> Fix for the `/* */` and `a.b` namespace issues is now in `master`.

for example, the resize operator has a "newHeight" attribute with format = "long". In the generated schema.js, it will be $.newHeight = reader.int64_(position, 8, 0). However, when I run the built Netron, it will report an error that "reader.int64_" is not a function in "*.ms"(the name of my model). The reason is that in flatbuffers.js, the class flatbuffers.Reader doesn't has a function named "int64_(position, offset, defaultValue)" .

**Comment by YvetteLaw** [comment]
> > > Can flatc.js process long(int64) data?
> > 
> > 
> > Can you provide an example?
> > Fix for the `/* */` and `a.b` namespace issues is now in `master`.
> 
> for example, the resize operator has a "newHeight" attribute with format = "long". In the generated schema.js, it will be $.newHeight = reader.int64_(position, 8, 0). However, when I run the built Netron, it will report an error that "reader.int64_" is not a function in "*.ms"(the name of my model). The reason is that in flatbuffers.js, the class flatbuffers.Reader doesn't has a function named "int64_(position, offset, defaultValue)" .

I have fixed this bug by add int64_()  function within class flatbuffers.Reader  in flatbuffers.js

File: source/view.js (hunk 1 above)
Comment (YvetteLaw) [reply source/view.js:1250]: In model.fbs, the “version” attribute of table MetaGraph can be used to determine whether a ’.ms' file is a MindSpore Lite model.

Also, in the line 38 of mslite.js, I've supplement a format matching judgement for the ".ms" file.

File: source/app.js (hunk 6 above)
Comment (YvetteLaw) [reply source/app.js:137]: Since the extension of our many models is .ms, changing it will have a great impact. Instead, in the line 38 of mslite.js, I've supplement a format matching to determine whether a ’.ms' file is a MindSpore Lite model.

**Comment by YvetteLaw** [comment]
Could you please merge my PR?

**Comment by lutzroeder** [comment]
```
~/Projects/netron: node ./test/models.js mslite
mslite/mobilenetv2.ms
Error: Invalid model format 'MindSpore Lite 1.0.0'.
    at ~/Projects/netron/test/models.js:494:19
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

File: source/view.js (hunk 1 above)
Comment (lutzroeder) [reply source/view.js:1250]: This does not work as you first need to successfully parse the FlatBuffers file to check the version. The file might not be a valid FlatBuffers file or in a different format that does not match the schema. The correct way to do this is using `file_identifier` and `file_extension`. Existing files might have to get recreated (great time to also fix the file extension from `.ms` to `.mslite`).

File: test/models.json [hunk 11]
```diff
@@ -5726,5 +5726,11 @@
     "target": "j48model.model",
     "source": "https://raw.githubusercontent.com/PTaati/wekaTree2python/master/j48model.model",
     "error":  "Unsupported type 'weka.classifiers.trees.J48' in 'j48model.model'."
+  },
+  {
+    "type":   "mslite",
```
Comment (lutzroeder) [inline test/models.json:5731]: `type` entries in `models.json` should be in alphabetical order.

**Review by YvetteLaw — COMMENTED** [review]
> ```
> ~/Projects/netron: node ./test/models.js mslite
> mslite/mobilenetv2.ms
> Error: Invalid model format 'MindSpore Lite 1.0.0'.
>     at ~/Projects/netron/test/models.js:494:19
>     at processTicksAndRejections (internal/process/task_queues.js:93:5)
> ```

The format of model has been changed to "MindSpore Lite 1.0.0" in models.json.

File: source/view.js (hunk 1 above)
Comment (YvetteLaw) [reply source/view.js:1250]: Since such changes will affect the compatibility of many current MindSpore Lite models, this PR needs to be suspended temporarily.

**Comment by YvetteLaw** [comment]
Some operators in our model have array-like attributes. Currently, there are the following two issues about the values of these attribute:
1. If an attribute is a high-dimensional array (ndim>1), the result displayed is a one-dimensional array. It is found that the shape information of the array is lost.
2. If an array is of type "long"(int64), the display result is wrong.

These two issues are what I'm trying to solve. Can you please help me?

File: source/flatbuffers.js [hunk 12]
```diff
@@ -103,6 +103,11 @@ flatbuffers.Reader = class {
         return this._dataView.getInt64(offset, true);
     }
 
+    int64_(position, offset, defaultValue) {
```
[the remark this answers is not in the document]
Comment (lutzroeder) [reply source/flatbuffers.js:106]: This is now added in `master`.

**Comment by lutzroeder** [comment]
> Some operators in our model have array-like attributes. Currently, there are the following two issues about the values of these attribute:
> 1. If an attribute is a high-dimensional array (ndim>1), the result displayed is a one-dimensional array. It is found that the shape information of the array is lost.
> 2. If an array is of type "long"(int64), the display result is wrong.
> These two issues are what I'm trying to solve. Can you please help me? 

Can you provide examples how to reproduce these issues?

**Comment by YvetteLaw** [comment]
> > Some operators in our model have array-like attributes. Currently, there are the following two issues about the values of these attribute:
> > 
> > 1. If an attribute is a high-dimensional array (ndim>1), the result displayed is a one-dimensional array. It is found that the shape information of the array is lost.
> > 2. If an array is of type "long"(int64), the display result is wrong.
> >    These two issues are what I'm trying to solve. Can you please help me?
> 
> Can you provide examples how to reproduce these issues?

You can download example models from https://pan.baidu.com/s/13jfs0eLhhDzyD1CpIFx1rw, the password is t7wa

1. In the model pad.ms, there is a node with type "Pad". The format of its "paddings" attribute is a 2-D array of "int"(int32). The value should be [[1, 1], [2, 2], [3, 3], [4, 4]] (seen from the second input in pad.tflite model). Hoeever, it is shown [1, 1, 2, 2, 3, 3, 4, 4] in Netron.

2. In the model reshape.ms, there is a node with type "Reshape". The format of  its "shape" attribute is a 1-D array of "long"(int64). The should be [3,5,20] (seen from the second input in reshape.tflite model). However, it is shown [3,0,5] in Netron.

**Comment by YvetteLaw** [comment]
Does Netron release a version every 7 days?

**Comment by lutzroeder** [comment]
> > 1. If an attribute is a high-dimensional array (ndim>1), the result displayed is a one-dimensional array. It is found that the shape information of the array is lost.
> > 2. If an array is of type "long"(int64), the display result is wrong.
> >    These two issues are what I'm trying to solve. Can you please help me?

1. The values are stored as a flat array and the app or `.fbs` have no knowledge of this encoding. It is up to the `mslite.Attribute` code to interpret the value if needed. For example, `-metadata.json` can include `"type": "int64[][2]"` and custom code in `mslite.Attribute` would transform the flat array into such a nested array.

2. is fixed in `master`.

File: source/flatbuffers.js (hunk 12 above)
Comment (lutzroeder) [inline source/flatbuffers.js:106]: Rebase to `master`.

File: source/mslite-metadata.json [hunk 13]
```diff
@@ -0,0 +1,2017 @@
...
+      "inputs": [
+        { "name": "x" }
+      ],
+      "outputs": [
+        { "name": "y" }
+      ]
+    }
+  },
+  {
+    "name": "Scale",
```
Comment (lutzroeder) [inline source/mslite-metadata.json:1598]: `Scale` is missing `"category": "Layer"`, `inputs` and `outputs`.

scale

File: source/mslite-metadata.json [hunk 14]
```diff
@@ -0,0 +1,2017 @@
...
+    "name": "SkipGram",
+    "schema": {
+      "inputs": [
+        { "name": "input"}
+      ],
+      "outputs": [
+        { "name": "output" }
+      ],
+      "attributes": [
+        { "name": "includeAllGrams", "type": "bool", "default": false },
```
Comment (lutzroeder) [inline source/mslite-metadata.json:1665]: Type names need to be consistent for all `-metadata.json` files. For example, use `boolean` instead of `bool`.

File: source/mslite-metadata.json [hunk 15]
```diff
@@ -0,0 +1,2017 @@
...
+        { "name": "maxSkipSize", "type": "int", "default": 0 },
```
Comment (lutzroeder) [inline source/mslite-metadata.json:1666]: `int32` or `int64` instead of `int`.

File: source/mslite-metadata.json [hunk 16]
```diff
@@ -0,0 +1,2017 @@
...
+      "category": "Tensor",
+      "inputs": [
+        { "name": "input"}
+      ],
+      "outputs": [
+        { "name": "output" }
+      ],
+      "attributes": [
+        { "name": "format", "type": "Format" },
+        { "name": "axes", "type": "[int]" },
```
Comment (lutzroeder) [inline source/mslite-metadata.json:1683]: `int32[]` or `int64[]` instead of `[int]`.

File: tools/mslite [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+
+set -e
+pushd $(cd $(dirname ${0})/..; pwd) > /dev/null
+
+bold() {
+    echo "$(tty -s && tput bold)$1$(tty -s && tput sgr0)"
+}
+
+clean() {
+    bold "mindspore clean"
```
Comment (lutzroeder) [inline tools/mslite:11]: `mslite clean`

File: tools/mslite [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+    rm -rf "./third_party/source/mindspore"
+}
+
+sync() {
+    bold "mindspore sync"
```
Comment (lutzroeder) [inline tools/mslite:16]: `mslite sync`

File: tools/mslite [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    mkdir -p "./third_party/source/mindspore/mindspore/lite/schema/"
+    curl --silent --location --output "./third_party/source/mindspore/mindspore/lite/schema/model.fbs" "https://github.com/mindspore-ai/mindspore/blob/master/mindspore/lite/schema/model.fbs?raw=true"
+    curl --silent --location --output "./third_party/source/mindspore/mindspore/lite/schema/ops.fbs" "https://github.com/mindspore-ai/mindspore/blob/master/mindspore/lite/schema/ops.fbs?raw=true"
+}
+
+schema() {
+    echo "mslite schema"
```
Comment (lutzroeder) [inline tools/mslite:23]: `mslite schema`

**Review by lutzroeder — CHANGES_REQUESTED** [review]

File: source/mslite-metadata.json [hunk 20]
```diff
@@ -0,0 +1,2140 @@
...
+      "inputs": [
+        { "name": "input" },
+        { "name": "shape" }
+      ],
+      "outputs": [
+        { "name": "output" }
+      ],
+      "attributes": [
+        { "name": "format", "type": "Format" },
+        { "name": "shape", "type": "int32[]" }
```
Comment (lutzroeder) [inline source/mslite-metadata.json:1578]: According to `ops.fbs` this should be `int64[]`? e.g. `long`=`int64` and `int`=`int32`.

Comment (YvetteLaw) [reply source/mslite-metadata.json:1867]: > According to `ops.fbs` this should be `int64[]`? e.g. `long`=`int64` and `int`=`int32`.

YES

**Comment by lutzroeder** [comment]
`Unsupported file content (18000000140018000000040000000800) for extension '.ms' in 'mobilenetv2.ms'.`

**Comment by YvetteLaw** [comment]
> ```
> ~ node ./test/models.js mslite
> mslite/mobilenetv2.ms
> Unsupported file content (18000000140018000000040000000800) for extension '.ms' in 'mobilenetv2.ms'.
> ```

I have updated the test models. Please verificate them again.

**Comment by YvetteLaw** [comment]
If our code can meet your requirements, please merge this PR   ^_^
